File: 9월 2일 - 정렬/2751.cpp
#include <iostream>
#include <vector>

using namespace std;
vector<int> arr, sorted;

void merge(int left, int middle, int right){
    int p1 = left; // 왼쪽 배열 포인터
    int p2 = middle + 1; // 오른쪽 배열 포인터
    int k = left; // 정렬된 배열의 포인터
    
    // 작은 순서대로 배열에 삽입
    while(p1 <= middle && p2 <= right){
        if(arr[p1] < arr[p2]){
            sorted[k++] = arr[p1++];
        }
        else{
            sorted[k++] = arr[p2++];
        }
    }

    // 남은 데이터 삽입
    while(p1 <= middle){
        sorted[k++] = arr[p1++];
    }
    while(p2 <= right){
        sorted[k++] = arr[p2++];
    }

    // 원래 배열에 복사
    for(int i = left; i <= right; i++){
        arr[i] = sorted[i];
    }
}

void mergeSort(int left, int right){
    if(left < right){ // 크기가 1보다 큰 경우만 고려
        int mid = (left+right)/2;
        mergeSort(left, mid);
        mergeSort(mid+1, right);

        merge(left, mid, right);
    }
}

int main(){
    int n;
    cin >> n;
    arr.assign(n, 0);
    sorted.assign(n, 0);

    for (int i=0; i<n; i++){
        cin >> arr[i];
    }

    mergeSort(0, n-1);

    for (int i=0; i<n; i++){
        cout << arr[i] << "\n";
    }
}

File: 9월 2일 - 정렬/Python/2751.py
# 정렬: 2751 수 정렬하기 2
import sys
input = sys.stdin.readline
n = int(input())
arr = []
for i in range(n):
    arr.append(int(input()))
arr.sort()
# arr = merge_sort(arr)
for i in range(len(arr)):
    print(arr[i])
"""
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    # 재귀함수를 이용해 끝까지 분할
    mid = len(arr)
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    i, j= 0, 0
    array = []
  
    # 둘중 하나조건에 부합하지 않을경우 while문 빠져나감 
    while i < len(left) and j < len(right): 
        if left[i] < right[j]:
            array.append(left[i])
            i+=1
        else:
            arr.append(right[j])
            j+=1

    # while문 빠져 나온 후, left혹은 right에 남은 요소들 arr에 넣어주기
    array += left[i:] 
    array += right[j:]
    
    return array"""